core/iwasm/fast-jit/cg/x86-64/jit_codegen_x86_64.cpp

```diff
@@ -6242,59 +6388,1271 @@ cast_r_f64_to_r_i64(x86::Assembler &a, int32 reg_no_dst, int32 reg_no_src)
+            else                                                           \
+                _ret = at_cmpxchg_imm_ra_base_r_offset_r(                  \
+                    a, bytes_dst, JIT_REG_KIND_##kind, &data_xchg,         \
+                    reg_no_base, reg_no_offset);                           \
+        }                                                                  \
+        else {                                                             \
+            if (jit_reg_is_const(r3))                                      \
+                _ret = at_cmpxchg_r_ra_base_r_offset_imm(                  \
+                    a, bytes_dst, JIT_REG_KIND_##kind, reg_no_xchg,        \
+                    reg_no_base, offset);                                  \
+            else                                                           \
+                _ret = at_cmpxchg_r_ra_base_r_offset_r(                    \
+                    a, bytes_dst, JIT_REG_KIND_##kind, reg_no_xchg,        \
+                    reg_no_base, reg_no_offset);                           \
+        }                                                                  \
+        if (!_ret)                                                         \
+            GOTO_FAIL;                                                     \
+    } while (0)
+
+/**
+ * Encode negate a value in the register
+ *
+ * @param a the assembler to emit the code
+ * @param bytes_dst the bytes number of the data,
+ *        could be 1(byte), 2(short), 4(int32), 8(int64),
+ * @param kind_dst the kind of data to move, could be I32, I64
+ * @param reg_no_src the index of register hold src value
+ *
+ * @return true if success, false otherwise
+ */
+static bool
+neg_r(x86::Assembler &a, uint32 bytes_dst, uint32 kind_dst, int32 reg_no_src)
+{
+    bh_assert((kind_dst == JIT_REG_KIND_I32 && bytes_dst <= 4)
+              || kind_dst == JIT_REG_KIND_I64);
+    bh_assert(reg_no_src < 16);
+    switch (bytes_dst) {
+        case 1:
+            a.neg(regs_i8[reg_no_src]);
+            break;
+        case 2:
+            a.neg(regs_i16[reg_no_src]);
+            break;
+        case 4:
+            a.neg(regs_i32[reg_no_src]);
+            break;
+        case 8:
+            a.neg(regs_i64[reg_no_src]);
+            break;
+        default:
+            bh_assert(0);
+            return false;
+    }
+    return true;
+}
+
+/**
+ * Encode atomic exchange and add
+ *
+ * @param a the assembler to emit the code
+ * @param bytes_dst the bytes number of the data,
+ *        could be 1(byte), 2(short), 4(int32), 8(int64),
+ * @param kind_dst the kind of data to move, could be I32, I64
+ * @param reg_no_dst the index of dest register
+ * @param reg_no_src the index of register hold operand value of add operation
+ * @param m_dst the dest memory operand
+ *
+ * @return true if success, false otherwise
+ */
+static bool
+at_xadd(x86::Assembler &a, uint32 bytes_dst, uint32 kind_dst, int32 reg_no_dst,
```

Comment:
Had better remove `int32 reg_no_dst`, it is unused in the function and makes confusing.